### PR TITLE
Kucoin: Margin subscription fix and improvements

### LINKED ISSUE #1755: Kucoin: Margin only-subs don't subscribe to spot pairs
As a GCT user with spot and margin assets enabled, but only margin asset enabled websocket subscriptions, I should still get subscriptions for all the pairs in margin which are also in spot

Currently it only works when spot subscriptions are enabled.
Otherwise the spot pairs are ignored.

### PR DESCRIPTION
**Pull request by gbjk**
* Fixes margin-only subscriptions not subscribing to the enabled spot asset pairs, despite there being no enabled spot sub
* All margin pairs are merged to the spot subscription, if there is one, reducing the number of subscriptions
* Require/Assert wording fixed in kucoin test

Fixes #1755

## Use case
As a GCT user with spot and margin assets enabled, but only margin asset enabled websocket subscriptions, I should still get subscriptions for all the pairs in margin which are also in spot Currently it only works when spot subscriptions are enabled. Otherwise the spot pairs are ignored.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## How has this been tested

- [x] go test ./... -race
- [x] golangci-lint run
- [x] `TestGenerateMarginSubscriptions`

### CODE REVIEW

**Review by shazbert — APPROVED** [review]
Thanks, nothing from me, tACK unauth.

File: exchanges/kucoin/kucoin_websocket.go [hunk 1]
```diff
@@ -1686,13 +1683,41 @@ func channelName(s *subscription.Subscription, a asset.Item) string {
 	return s.Channel
 }
 
-// removeSpotFromMargin removes spot pairs from margin pairs in the supplied AssetPairs map for subscriptions to non-margin endpoints
-func removeSpotFromMargin(s *subscription.Subscription, ap map[asset.Item]currency.Pairs, spotPairs currency.Pairs) string {
+// mergeMarginPairs merges margin pairs into spot pairs for shared subs (ticker, orderbook, etc) if Spot asset and sub are enabled,
+// because Kucoin errors on duplicate pairs in separate subs, and doesn't have separate subs for spot and margin
+func (ku *Kucoin) mergeMarginPairs(s *subscription.Subscription, ap map[asset.Item]currency.Pairs) string {
```
Comment (gloriousCode) [inline exchanges/kucoin/kucoin_websocket.go:1688]:
```
[ERROR] | WEBSOCKET | 31/12/2024 09:57:40 | Kucoin websocket: subscription failure, ticker all : subscription template did not generate the expected number of asset records: Got 2; Expected 1, orderbook all : subscription template did not generate the expected number of asset records: Got 2; Expected 1
```
I hadn't even tried to mess with things yet. I accidentally had margin asset pair disabled 😭 

Config gist: [here](https://gist.github.com/gloriousCode/8b80a088d462213b40f3b68ddba4a694)
Have I done something wrong?
edit: works when I enable margin...

**Review by gloriousCode — CHANGES_REQUESTED** [review]

File: exchanges/kucoin/kucoin_websocket.go [hunk 2]
```diff
@@ -1686,13 +1683,41 @@ func channelName(s *subscription.Subscription, a asset.Item) string {
...
 	if strings.HasPrefix(s.Channel, "/margin") {
 		return ""
 	}
-	if p, ok := ap[asset.Margin]; ok {
-		ap[asset.Margin] = p.Remove(spotPairs...)
+	wantKey := &subscription.IgnoringAssetKey{Subscription: s}
+	switch s.Asset {
+	case asset.All:
+		marginPairs, _ := ku.GetEnabledPairs(asset.Margin)
+		ap[asset.Spot] = common.SortStrings(ap[asset.Spot].Add(marginPairs...))
```
[the remark this answers is not in the document]
Comment (gbjk) [reply exchanges/kucoin/kucoin_websocket.go:1696]: Only important for consistency and testing.
They're sorted originally in `fillAssetPairs`. 
Adding to them later implies they should remain sorted.
The reason it matters is because the sequence is what we'll actually use in the QualifiedChannel.
I subjectively wanted that to be consistent and deterministic, but maybe it doesn't have to be.

If we don't sort them here, we'll have to do some magic in testing things like QualifiedChannel batching into sets of 100, etc. Off the top of my head, I think we'd have to lose the test helper for equalSubs if we don't have a deterministic QualifiedChannel, because the helper can't know how to split it, especially for batching pairs.

I'm not advocating runtime performance for testing convenience. The bigger argument for me is subscription message consistency.

Without sorting, this is the failure we get for kucoin:
```
    kucoin_test.go:2295:
                Error Trace:    /Users/gbjk/go/src/github.com/gbjk/gocryptotrader/internal/testing/subscriptions/subscriptions.go:32
                                                        /Users/gbjk/go/src/github.com/gbjk/gocryptotrader/exchanges/kucoin/kucoin_test.go:2295
                Error:          Differences:
                                 + /market/ticker:BTC-USDT,ETH-BTC,ETH-USDT,LTC-USDT,TRX-BTC,SOL-USDC ticker spot BTC/USDT,ETH/BTC,ETH/USDT,LTC/USDT,TRX/BTC,SOL/USDC
                                 + /spotMarket/level2Depth5:BTC-USDT,ETH-BTC,ETH-USDT,LTC-USDT,TRX-BTC,SOL-USDC orderbook spot BTC/USDT,ETH/BTC,ETH/USDT,LTC/USDT,TRX/BTC,SOL/USDC
                                 - /market/ticker:BTC-USDT,ETH-BTC,ETH-USDT,LTC-USDT,SOL-USDC,TRX-BTC ticker spot BTC/USDT,ETH/BTC,ETH/USDT,LTC/USDT,SOL/USDC,TRX/BTC
                                 - /spotMarket/level2Depth5:BTC-USDT,ETH-BTC,ETH-USDT,LTC-USDT,SOL-USDC,TRX-BTC orderbook spot BTC/USDT,ETH/BTC,ETH/USDT,LTC/USDT,SOL/USDC,TRX/BTC
                Test:           TestGenerateSubscriptions
                Messages:       Subscriptions should be equal
```

File: exchanges/kucoin/kucoin_websocket.go (hunk 1 above)
Comment (gbjk) [reply exchanges/kucoin/kucoin_websocket.go:1688]: Thanks so much for catching that 🙏  Added a test case.
Need to ensure we don't add an empty margin asset which didn't exist.
Kucoin remains the place we're fangling assetPairs the most, so this is quite edge-casey.

Fixed https://github.com/gbjk/gocryptotrader/commit/a7e1f47fa

Comment (gbjk) [reply exchanges/kucoin/kucoin_websocket.go:1688]: I've seen another bug now when margin is all that's enabled. I'm on it.

Comment (gbjk) [reply exchanges/kucoin/kucoin_websocket.go:1688]: Got it 🎉  Just need to leave assets alone in `asset.All` subs unless there's *both* margin and spot subs available.

Fixed https://github.com/gbjk/gocryptotrader/commit/d4ac07b70

File: exchanges/kucoin/kucoin_websocket.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (gloriousCode) [reply exchanges/kucoin/kucoin_websocket.go]: Lots of stuff there but as an overall rule "sub currencies are sorted" is reasonable

**Review by gloriousCode — APPROVED** [review]
tACK!

**Review by shazbert — APPROVED** [review]
ch ACK.

**Review by thrasher- — APPROVED** [review]
tACK margin only, spot/margin with same pairs, spot/margin with different pairs and all looks to work great
